pchrt.1
.\"                                      Hey, EMACS: -*- nroff -*-
.TH PCHRT 1 "May 10 2016"
.SH NAME
.B pchrt (python\-schedutils) \- manipulate the real-time attributes of a process
.SH SYNOPSIS
pchrt [options] [prio] [pid | cmd [args...]]
.SH OPTIONS
.TP
.B  \-b, \-\-batch
set policy to SCHED_BATCH
.TP
.B  \-f, \-\-fifo
set policy to SCHED_FIFO
.TP
.B  \-i, \-\-idle
set policy to SCHED_IDLE
.TP
.B  \-p, \-\-pid
operate on existing given pid
.TP
.B  \-m, \-\-max
show min and max valid priorities
.TP
.B  \-o, \-\-other
set policy to SCHED_OTHER
.TP
.B  \-r, \-\-rr
set policy to SCHED_RR (default)
.TP
.B  \-R, \-\-reset-on-fork
set SCHED_RESET_ON_FORK for FIFO or RR
.TP
.B  \-h, \-\-help
display help
.TP
.B Note:
You must give a priority if changing policy.
.SH SEE ALSO
.BR chrt (1)
.SH AUTHOR
pchrt was written by Arnaldo Carvalho de Melo <acme@redhat.com>
.br
as part of python\-schedutils
.PP
This manual page was written by John Kacur <jkacur@redhat.com>
.PP
Report bugs and send patches to <tuna-devel@lists.fedorahosted.org>


ChangeLog
v0.4:

[acme@mica python-schedutils]$ git log v0.2.. --format=oneline --abbrev-commit
1568357 Fix cpusetsize parameter in sched_setaffinity call
28a2f17 Support SCHED_IDLE in schedfromstr
ac29a2e Use PyErr_SetFromErrno return value
f09bae5 Fix leak on setaffinity error path
76269f0 Don't set the error string twice
80ce42f Don't limit looking for CPUs at 1024
028a3af Use PyErr_NoMemory when CPU_ALLOC fails
36e377d CPU_ALLOC macros are not present in older systems
6b9f411 sched_setaffinity expects a cpu_set_t pointer, not a ptr to a pointer
741697a Dinamically size the cpu_set_t objects
4f3b33c Support SCHED_RESET_ON_FORK policy flag
66a2327 Add SCHED_IDLE to the list of constants exported
4e05f75 Methods should raise OSError exceptions
d3aead2 schedutils: Older kernels don't have SCHED_BATCH defined
c162a1e Merge branch 'master' of ssh://master.kernel.org/pub/scm/linux/kernel/git/acme/python-schedutils
fb72c8e Makefile: fixup version gathering
10f78ff rpm: Fixed %build and %install sections, part of fedora review


pchrt.py
#! /usr/bin/python
# -*- python -*-
# -*- coding: utf-8 -*-
#   Copyright (C) 2008 Red Hat Inc.
#
#   Arnaldo Carvalho de Melo <acme@redhat.com>
#
#   This application is free software; you can redistribute it and/or
#   modify it under the terms of the GNU General Public License
#   as published by the Free Software Foundation; version 2.
#
#   This application is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#   General Public License for more details.

import os, schedutils, sys

def usage():
	print '''pchrt (python-schedutils)
usage: pchrt [options] [prio] [pid | cmd [args...]]
manipulate real-time attributes of a process
  -b, --batch                        set policy to SCHED_BATCH
  -f, --fifo                         set policy to SCHED_FIFO
  -i, --idle                         set policy to SCHED_IDLE
  -p, --pid                          operate on existing given pid
  -m, --max                          show min and max valid priorities
  -o, --other                        set policy to SCHED_OTHER
  -r, --rr                           set policy to SCHED_RR (default)
  -R, --reset-on-fork                set SCHED_RESET_ON_FORK for FIFO or RR
  -h, --help                         display this help

You must give a priority if changing policy.

Report bugs and send patches to <tuna-devel@lists.fedorahosted.org>'''
	return

def show_priority_limits(policy):
	print "%-32.32s: %d/%d" % ("%s min/max priority" % schedutils.schedstr(policy),
				   schedutils.get_priority_min(policy),
				   schedutils.get_priority_max(policy))

def show_all_priority_limits():
	for policy in (schedutils.SCHED_OTHER, schedutils.SCHED_FIFO,
		       schedutils.SCHED_RR, schedutils.SCHED_BATCH):
		show_priority_limits(policy)

def show_settings(pid):
	policy = schedutils.get_scheduler(pid)
	spolicy = schedutils.schedstr(policy)
	rtprio = schedutils.get_priority(pid)
	reset_on_fork = ""
	if policy & schedutils.SCHED_RESET_ON_FORK:
		reset_on_fork = "|SCHED_RESET_ON_FORK"
	print '''pid %d's current scheduling policy: %s%s
pid %d's current scheduling priority: %d''' % (pid, spolicy, reset_on_fork, pid, rtprio)

def valid_policy_flag(policy, policy_flag):
	if policy_flag == schedutils.SCHED_RESET_ON_FORK and \
	   policy not in (schedutils.SCHED_RR, schedutils.SCHED_FIFO):
		print "SCHED_RESET_ON_FORK flag is supported for SCHED_FIFO and SCHED_RR policies only"
		return False
	return True

def change_settings(pid, policy, policy_flag, rtprio):
	try:
		schedutils.set_scheduler(pid, policy | policy_flag, rtprio)
	except SystemError, err:
		print "sched_setscheduler: %s" % err[1]
		print "failed to set pid %d's policy" % pid

def main():

	args = sys.argv[1:]
	if not args:
		usage()
		return

	policy = schedutils.SCHED_RR
	policy_flag = 0
	while True:
		o = args.pop(0)
		try:
			priority = int(o)
			break
		except:
			pass

		if o in ("-h", "--help"):
			usage()
			return
		elif o in ("-b", "--batch"):
			policy = schedutils.SCHED_BATCH
		elif o in ("-f", "--fifo"):
			policy = schedutils.SCHED_FIFO
		elif o in ("-i", "--idle"):
			policy = schedutils.SCHED_IDLE
		elif o in ("-m", "--max"):
			show_all_priority_limits()
			return
		elif o in ("-o", "--other"):
			policy = schedutils.SCHED_OTHER
		elif o in ("-r", "--rr"):
			policy = schedutils.SCHED_RR
		elif o in ("-R", "--reset-on-fork"):
			policy_flag |= schedutils.SCHED_RESET_ON_FORK
		elif o in ("-p", "--pid"):
			if len(args) > 1:
				priority = int(args.pop(0))
				pid = int(args.pop(0))
				if not valid_policy_flag(policy, policy_flag):
					return
				change_settings(pid, policy, policy_flag, priority)
			else:
				pid = int(args.pop(0))
				show_settings(pid)
			return
		else:
			usage()
			return

	if not valid_policy_flag(policy, policy_flag):
		return

	schedutils.set_scheduler(0, policy | policy_flag, priority)
	os.execvp(args[0], args)

if __name__ == '__main__':
    main()
